doc/sphinx/language/core/assumptions.rst

```diff
@@ -71,7 +71,21 @@ the list of binders, it is expanded to a let-in definition (see
 Section :ref:`let-in`).
 
 The expression :n:`forall @ident : @type, @term` denotes the
-*product* of the variable :n:`@ident` of type :n:`@type`, over the term :n:`@term`.
+:gdef:`product type <product>` (or *product*) of the variable :n:`@ident` of type :n:`@type`,
+over the term :n:`@term`.  If :n:`@ident` is used in :n:`@term`, then we say
+the expression is a :gdef:`dependent product`.
+
+These terms are also useful:
+
+* `n` is a :gdef:`dependent premise` of `forall n:nat, n + 0 = n` because
+  `n` appears both in the binder of the `forall` and in the quantified statement
+  `n + 0 = n`.
+
+* `A` and `B` are :gdef:`non-dependent premises <non-dependent premise>` of `A -> B -> C`
+  because they don't appear in a `forall` binder.
+  (Recall that `A -> B` is notation for the term `forall _ : A, B`; the wildcard
+  `_` can't be referred to in the quantified statement.)
```

Comment:
There is a problem of order with this "Recall" which is that this notation is explained at the very bottom of this section.

Reply:
I moved it, please take a look.

Are you OK with dropping the last paragraph (which I don't fully understand)?:

![image](https://user-images.githubusercontent.com/1253341/122469246-8f751080-cf71-11eb-8766-2697ea14bfd2.png)